File: 01-05-2024/Arrange Consonants and Vowels.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

struct Node {
    char data;
    struct Node *next;

    Node(int x) {
        data = x;
        next = NULL;
    }
};

void printlist(Node *head) {
    if (head == NULL)
        return;
    while (head != NULL) {
        cout << head->data << " ";
        head = head->next;
    }
    cout << endl;
}

void append(struct Node **headRef, char data) {
    struct Node *new_node = new Node(data);
    struct Node *last = *headRef;

    if (*headRef == NULL) {
        *headRef = new_node;
        return;
    }
    while (last->next != NULL)
        last = last->next;
    last->next = new_node;
    return;
}


// } Driver Code Ends
/*
Structure of the node of the linked list is as
struct Node
{
    char data;
    struct Node *next;

    Node(char x){
        data = x;
        next = NULL;
    }

};
*/
// Your Code Start
class Solution {
public:
    Node* arrangeCV(Node* head) {
        vector<char> vowels = {'a', 'e', 'i', 'o', 'u'};
        Node* dummy = new Node(-1);
        dummy->next = head;
        Node* vow = dummy;
        Node* prev = dummy;
        Node* curr = head;
        
        while (curr) {
            if (find(vowels.begin(), vowels.end(), curr->data) != vowels.end()) {
                Node* del_node = curr;
                prev->next = curr->next;
                curr = prev->next;
                Node* nxt = vow->next;
                vow->next = del_node;
                del_node->next = nxt;
                vow = del_node;
                if (vow->next == curr) {
                    prev = vow;
                }
            } else {
                curr = curr->next;
                prev = prev->next;
            }
        }
        return dummy->next;
    }
};
//Your Code End


//{ Driver Code Starts.

// task is to complete this function
int main() {
    int T;
    cin >> T;
    while (T--) {
        int n;
        char tmp;
        struct Node *head = NULL;
        cin >> n;
        while (n--) {
            cin >> tmp;
            append(&head, tmp);
        }
        Solution obj;
        head = obj.arrangeCV(head);
        printlist(head);
    }
    return 0;
}

// } Driver Code Ends

File: 01-05-2024/Arrange Consonants and Vowels.js
//{ Driver Code Starts
//Initial Template for javascript

"use strict";

process.stdin.resume();
process.stdin.setEncoding("utf-8");

let inputString = "";
let currentLine = 0;



process.stdin.on("data", (inputStdin) => {
  inputString += inputStdin;
});

process.stdin.on("end", (_) => {
  inputString = inputString
    .trim()
    .split("\n")
    .map((string) => {
      return string.trim();
    });

  main();
});

function readLine() {
  return inputString[currentLine++];
}

class Node {
  constructor(x){
    this.data = x;
    this.next = null;
  }
}

function printList(head){
  let s = "";
  while(head){
    s+=head.data;
    s+=" ";
    head = head.next;
  }
  console.log(s);
}

let a = null;
let b = null;

function main() {
  let t = parseInt(readLine());
  let i = 0;
  for (; i < t; i++) {
    let n = parseInt(readLine());
    let arr = readLine().trim().split(" ");
    let list = new Node(arr[0]);
    let tail = list;
    for(let j = 1;j<n;j++){
      tail.next = new Node(arr[j]);
      tail = tail.next;
    }
    
    let obj = new Solution();
    let res = obj.arrangeCV(list);
    printList(res);
  } 

}
// } Driver Code Ends


//User function Template for javascript

/*LINKED LIST NODE
class Node {
  constructor(x){
    this.data = x;
    this.next = null;
  }
}
*/

/**
 * @param {Node} head
 * @return {Node}
*/
//your Code Start

class Solution {
  arrangeCV(head) {
    if (!head || !head.next) return head; // If the list is empty or has only one node, return as it is
    
    let vowelHead = null, vowelTail = null;
    let consonantHead = null, consonantTail = null;
    
    let current = head;
    while (current) {
      if ('aeiou'.includes(current.data.toLowerCase())) {
        // If the current node is a vowel
        if (!vowelHead) {
          vowelHead = vowelTail = current;
        } else {
          vowelTail.next = current;
          vowelTail = current;
        }
      } else {
        // If the current node is a consonant
        if (!consonantHead) {
          consonantHead = consonantTail = current;
        } else {
          consonantTail.next = current;
          consonantTail = current;
        }
      }
      current = current.next;
    }
    
    // Connect the vowel list with the consonant list
    if (vowelHead) {
      vowelTail.next = consonantHead;
    }
    
    // Set the next of the last node in consonant list to null
    if (consonantTail) {
      consonantTail.next = null;
    }
    
    // Return the head of the rearranged linked list
    return vowelHead || consonantHead;
  }
}
//Your Code End
